server/src/test/java/org/elasticsearch/cluster/coordination/PeerFinderTests.java

```diff
@@ -0,0 +1,671 @@
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.nullValue;
+
+public class PeerFinderTests extends ESTestCase {
+
+    private CapturingTransport capturingTransport;
+    private DeterministicTaskQueue deterministicTaskQueue;
+    private DiscoveryNode localNode;
+    private MockUnicastHostsProvider unicastHostsProvider;
+    private MockTransportAddressConnector transportAddressConnector;
+    private TestPeerFinder peerFinder;
+
+    private Set<DiscoveryNode> disconnectedNodes = new HashSet<>();
+    private Set<DiscoveryNode> connectedNodes = new HashSet<>();
+    private DiscoveryNodes lastAcceptedNodes;
+    private TransportService transportService;
+
+    private static long CONNECTION_TIMEOUT_MILLIS = 30000;
+
+    class MockTransportAddressConnector implements TransportAddressConnector {
+        final Set<DiscoveryNode> reachableNodes = new HashSet<>();
+        final Set<TransportAddress> unreachableAddresses = new HashSet<>();
+        final Set<TransportAddress> slowAddresses = new HashSet<>();
+        final Set<TransportAddress> inFlightConnectionAttempts = new HashSet<>();
+
+        @Override
+        public void connectToRemoteMasterNode(TransportAddress transportAddress, ActionListener<DiscoveryNode> listener) {
+            assert localNode.getAddress().equals(transportAddress) == false : "should not probe local node";
+
+            final boolean isNotInFlight = inFlightConnectionAttempts.add(transportAddress);
+            assertTrue(isNotInFlight);
+
+            final long connectResultTime = deterministicTaskQueue.getCurrentTimeMillis()
+                + (slowAddresses.contains(transportAddress) ? CONNECTION_TIMEOUT_MILLIS : 0);
+
+            deterministicTaskQueue.scheduleAt(connectResultTime, new Runnable() {
+                @Override
+                public void run() {
+                    if (unreachableAddresses.contains(transportAddress)) {
+                        assertTrue(inFlightConnectionAttempts.remove(transportAddress));
+                        listener.onFailure(new IOException("cannot connect to " + transportAddress));
+                        return;
+                    }
+
+                    for (final DiscoveryNode discoveryNode : reachableNodes) {
+                        if (discoveryNode.getAddress().equals(transportAddress)) {
+                            if (discoveryNode.isMasterNode()) {
+                                disconnectedNodes.remove(discoveryNode);
+                                connectedNodes.add(discoveryNode);
+                                assertTrue(inFlightConnectionAttempts.remove(transportAddress));
+                                listener.onResponse(discoveryNode);
+                                return;
+                            } else {
+                                listener.onFailure(new ElasticsearchException("non-master node " + discoveryNode)); // TODO better exception
```

Comment:
Does this need a better exception or is this ok?

Reply:
for the tests, I think it doesn't matter